internal/handlers/common/saslstart.go

```diff
@@ -15,54 +15,42 @@
 package common
 
 import (
-	"bytes"
-	"encoding/base64"
+	"context"
 	"fmt"
 
+	"github.com/FerretDB/FerretDB/internal/clientconn/conninfo"
 	"github.com/FerretDB/FerretDB/internal/handlers/commonerrors"
 	"github.com/FerretDB/FerretDB/internal/types"
 	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
 )
 
-// SASLStartPlain extracts username and password from PLAIN `saslStart` payload.
-func SASLStartPlain(doc *types.Document) (string, string, error) {
-	var payload []byte
-
-	// some drivers send payload as a string
-	stringPayload, err := GetRequiredParam[string](doc, "payload")
-	if err == nil {
-		if payload, err = base64.StdEncoding.DecodeString(stringPayload); err != nil {
-			return "", "", lazyerrors.Error(err)
-		}
-	}
-
-	// most drivers follow spec and send payload as a binary
-	binaryPayload, err := GetRequiredParam[types.Binary](doc, "payload")
-	if err == nil {
-		payload = binaryPayload.B
+// SASLStart is a common implementation of the saslStart command.
+func SASLStart(ctx context.Context, doc *types.Document) error {
+	mechanism, err := GetRequiredParam[string](doc, "mechanism")
+	if err != nil {
+		return lazyerrors.Error(err)
 	}
 
-	if payload == nil {
-		// return error about expected types.Binary, not string
-		return "", "", err
-	}
-
-	parts := bytes.Split(payload, []byte{0})
-	if l := len(parts); l != 3 {
-		return "", "", commonerrors.NewCommandErrorMsgWithArgument(
-			commonerrors.ErrTypeMismatch,
-			fmt.Sprintf("Invalid payload (expected 3 parts, got %d)", l),
-			"payload",
+	var username, password string
+
+	switch mechanism {
+	case "PLAIN":
+		username, password, err = saslStartPlain(doc)
+	default:
+		msg := fmt.Sprintf("Unsupported authentication mechanism %q.\n", mechanism) +
+			"See https://docs.ferretdb.io/security/authentication/ for more details."
+		err = commonerrors.NewCommandErrorMsgWithArgument(
```

Comment:
We should return error here within default case instead of outside the `switch`

Reply:
fixed